File: queue.c
#include "stdio.h"
#include "stdlib.h"
#include "stdbool.h"
#include "queue.h"

bool enqueue(Queue **Q, Node **node)
{
	if (Q == NULL || (*Q) == NULL)
		return false;

	QNode *newElement = malloc(sizeof(QNode));
	if (newElement == NULL)
		return false;

	newElement->node = (*node);
	newElement->next = (*Q)->head;
	newElement->prev = NULL;

	if ((*Q)->head != NULL)
		(*Q)->head->prev = newElement;

	(*Q)->head = newElement;

	if ((*Q)->tail == NULL)
		(*Q)->tail = (*Q)->head;

	return true;
}

bool dequeue(Queue **Q, Node **node)
{
	if (Q == NULL || (*Q) == NULL)
		return false;

	if (isEmpty(*Q))
		return false;
	
	QNode *front = (*Q)->tail;
	
	if (front != NULL)
	{
		(*node) = front->node;
		free(front);
	}

	if ((*Q)->tail != NULL)
		(*Q)->tail = (*Q)->tail->prev;


	if ((*Q)->tail == NULL)
		(*Q)->head = NULL;

	return true;
}

bool initialize(Queue **Q)
{
	(*Q) = malloc(sizeof(Queue));
	(*Q)->head = NULL;
	(*Q)->tail = NULL;
}

bool destroy(Queue **Q)
{
	if (Q != NULL && (*Q) != NULL)
	{
		free(*Q);
	}
}

bool isEmpty(Queue *Q)
{
	return Q == NULL || Q->head == NULL || Q->tail == NULL;
}

File: Queue.java
public class Queue {
	class Node {
		Integer item;
		Node next;
		Node prev;
	}

	Node front;
	Node rear;

	public void enqueue(Integer item) {
		Node node = new Node();
		if (node == null)
			throw new RuntimeException("Queue is full");

		node.item = item;
		node.next = front;
		
		if (front != null)
			front.prev = node;
		front = node;

		if (rear == null)
			rear = front;
	}

	public Integer dequeue() {
		if (rear == null)
			throw new RuntimeException("Queue is Empty");

		Node node = rear;
		rear = rear.prev;

		if (rear == null)
			front = null;

		return node.item;
	}

	public static void main(String[] args) {
		Queue q = new Queue();
		q.enqueue(1);
		q.enqueue(2);
		q.enqueue(3);

		System.out.println(q.dequeue());
		System.out.println(q.dequeue());
		System.out.println(q.dequeue());
		System.out.println(q.dequeue());
	}
}
